crates/typst/src/text/mod.rs

```diff
@@ -1165,3 +1200,44 @@ impl Fold for WeightDelta {
         Self(outer.0 + self.0)
     }
 }
+
+/// Costs that are updated (prioritizing the later value) when folded.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Hash)]
+#[non_exhaustive] // We may add more costs in the future.
+pub struct Costs {
+    pub hyphenation: Smart<Ratio>,
+    pub runt: Smart<Ratio>,
+    pub widow: Smart<Ratio>,
+    pub orphan: Smart<Ratio>,
+}
+
+impl Fold for Costs {
+    fn fold(self, other: Self) -> Self {
+        Self {
+            hyphenation: other.hyphenation.or(self.hyphenation),
+            runt: other.runt.or(self.runt),
+            widow: other.widow.or(self.widow),
+            orphan: other.orphan.or(self.orphan),
+        }
+    }
+}
+
+cast! {
+    Costs,
+    self => dict![
+        "hyphenation" => self.hyphenation,
+        "runt" => self.runt,
+        "widow" => self.widow,
+        "orphan" => self.orphan,
+    ].into_value(),
+    mut v: Dict => {
+        let ret = Self {
+            hyphenation: v.take("hyphenation").ok().unwrap_or(Value::Auto).cast()?,
```

Comment:
I think we shouldn't expose `auto` to the user at all and only use it internally. Maybe `Option` is more reasonable then after all. That's how `Sides<Option<T>>` works. The `IntoValue` (i.e. `self => ..`) would then only add a field if present.

For `Stroke`, it's solved differently and the `auto` is exposed, but it's also a proper type, which complicates things. It's also not necessarily ideal there. 

For the `costs`, we could then also have the `Default` impl return `Some(100%)` rather than `None`, such that `context text.costs` yields a dictionary of ratios rather than `autos`.

If we do these changes, one sentence in the docs needs updating.

Reply:
I agree, but I think it might make more sense to always include all of the values, using 100% if the value was never specified.

Reply:
but then we can't override just some of them in multiple rules, e.g.?
```
#set text(costs: (hyphenation: 600%))
#set text(costs: (orphan: 200%))
```
If I understand you correctly, `hyphenation` would be back at 100% in your proposal.

Reply:
I meant when Costs is converted to a value, not vice versa. Your example works correctly. I think the only case where this shows up is `#context text.costs`, which I am proposing should contain all the resolved costs, which will be 100% for those that have not been set. I think this would be good because it avoids special cases when people want to get the value of a cost, e.g. `#context text.costs.orphan`, instead of having to check for the field presence and use a (hard-coded magic number) default value if not.

Reply:
Sorry, can you explain further what you are suggesting?

Reply:
OK, I understand your point, but I don't see what you intend to accomplish or what problem you see with the current approach.

Reply:
My problem is that users can write `#set text(costs: (hyphenation: auto))`. I don't want to expose `auto` to the users since no smart default is at play. The fact that the dictionary can be partially populated is purely an implementation detail. For this reason, I think an unexposed `Option` that only reflects the presence in the dictionary rather than an explicit `none` makes more sense.

Reply:
Okay, nice. I was still looking at the old code snippet I think.

Reply:
Ah yeah, have to be careful about that:

![Indicating "Outdated" on the change that was reviewed](https://github.com/typst/typst/assets/58113890/0c33ffce-d538-46a7-b3de-629ce52eca41)